web/src/main/webapp/WEB-INF/config-security/config-security-openidconnectbearer-overrides.properties

```diff
@@ -0,0 +1,96 @@
+# Copyright (C) 2020 Food and Agriculture Organization of the
+# United Nations (FAO-UN), United Nations World Food Programme (WFP)
+# and United Nations Environment Programme (UNEP)
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or (at
+# your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+#
+# Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
+# Rome - Italy. email: geonetwork@osgeo.org
+
+
+# SETUP
+# 1.
+
+#you MUST set these env vars;
+#----------------------------
+#
+# OPENIDCONNECT_CLIENTID   -- client id (from your IDP server)
+# OPENIDCONNECT_CLIENTSECRET  -- client secret (from your IDP server)
+#
+# IDP server metadata configuration is available from your IDP server;
+#     keycloak: http://localhost:8080/realms/{realm}/.well-known/openid-configuration
+#     azure AD: https://login.microsoftonline.com/00000000-0000-0000-0000-000000000000/v2.0/.well-known/openid-configuration
+#               where 00000000-0000-0000-0000-000000000000 is the directory (tenant) ID
+# OPENIDCONNECT_SERVERMETADATA_FNAME  -- file location for the IDP server metadata file - i.e. /WEB-INF/config-security/openid-configuration.json
+# Alternatively, you can put the actual text of the .json file in the environment variable;
+#  OPENIDCONNECT_SERVERMETADATA_JSON_TEXT
+#  NOTE: if you specify both OPENIDCONNECT_SERVERMETADATA_FNAME and OPENIDCONNECT_SERVERMETADATA_JSON_TEXT
+#        then OPENIDCONNECT_SERVERMETADATA_JSON_TEXT will be used!
```

Comment:
Wondering why  OPENIDCONNECT_SERVERMETADATA_JSON_TEXT was added?  Seems odd to create a variable with the results from .well-known/openid-configuration

I would expect it to work similar to the spring
i.e. they have a variable similar to the following.

#Sample keycloak
spring.security.oauth2.client.provider.oidc.issuer-uri=http://localhost:8080/realms/{realm}
#For Azure it would be the following
spring.security.oauth2.client.provider.oidc.issuer-uri=https://login.microsoftonline.com/00000000-0000-0000-0000-000000000000/v2.0

It would automatically download the json by appending "/.well-known/openid-configuration" to the url and making the http request.

This would be easier than manually downloading the json file.

Same comment as for config-security-openidconnect-overrides.properties file as well.

Reply:
I didn't want to have the startup of Geonetwork rely on an External HTTP service during startup.  Also, I didn't like the idea of the security configuration being hosted outside of GN.  
I, also, wasn't sure if all IDP server would have this file (although most do, but I didn't see anything in the spec saying it was required - but I could have missed that).
I, also, found that some of the JSON text wasn't necessary correct and might need to be modified - I didn't actually see a need for this.

Its in an environment variable to be easy for docker-ization.

It would be easy to add downloading the URL.

Reply:
As the OIDC server needs to be accessible by the geonetwork application to do authentication, geonetwork will need access to the oidc server so it makes sense for it to query the .well-known/openid-configuration url.

According to the following it is well documented and required for providers supporting Discovery.  If they don't support Discovery then I don't know where the json data will come from?

https://openid.net/specs/openid-connect-discovery-1_0.html
> 
> 
> OpenID Providers supporting Discovery MUST make a JSON document available at the path formed by concatenating the string /.well-known/openid-configuration to the Issuer. The syntax and semantics of .well-known are defined in [RFC 5785](https://openid.net/specs/openid-connect-discovery-1_0.html#RFC5785) [RFC5785] and apply to the Issuer value when it contains no path component. openid-configuration MUST point to a JSON document compliant with this specification and MUST be returned using the application/json content type.

I also find it odd that the spring setup just uses the url and I would imagine that it is capable of recovering from the oidc being down at startup (I don't know for sure).  So I find it odd that you cannot link to the same logic?

If you need to do the check/reload manually then I understand that it may be difficult to handle cases where the oidc server is not running at the time geonetwork is started.  I'm sure it is possible to check the authentication service status and keep trying to get the json file each time authentication requests are made.  If you cannot get the json file and the oidc server is down then authentication will not work anyway.  

At the very least, we can let the administrator decide to use this option. We can put a comment about the issue in the documentation for now and someday, someone can add the enhancement to keep trying to download the file on each request in a future PR.  

So far, this is the first application that I have seen where it requires the administrator to manually download and install the json file - all other application that I have used that connect to an oidc (The list is not that long) only require part of the url in order to derive(or direct link) the location of .well-known/openid-configuration.

Personally I would use the url and live with the issue of the oidc server being down at startup.

Reply:
1. For keycloak, when you configure an OIDC source, you manually press a button to download the JSON and then configure manually.  It only downloads the configuration once (manually), not at each startup.
2. Same for Geoserver

What systems have you seen that downloads the configuration at every startup?

I can add that functionality...

Reply:
You may be right that it was only downloading it once - I'm not sure what happened in the back end?

If you look at the following 
https://docs.spring.io/spring-security/site/docs/5.2.12.RELEASE/reference/html/oauth2.html#oauth2login-boot-property-mappings

> --
> A ClientRegistration can be initially configured using discovery of an OpenID Connect Provider’s Configuration endpoint or an Authorization Server’s Metadata endpoint, by specifying the spring.security.oauth2.client.provider.[providerId].issuer-uri property.
> 
> [Tip]
> A ClientRegistration can be initially configured using discovery of an OpenID Connect Provider’s [Configuration endpoint](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig) or an Authorization Server’s [Metadata endpoint](https://tools.ietf.org/html/rfc8414#section-3), by specifying the spring.security.oauth2.client.provider.[providerId].issuer-uri property.

This is similar to the `spring.security.oauth2.client.provider.oidc.issuer-uri` that I mentioned above.  I assumed that if it was done in the properties then at every start it would use this to get all the required information from the authentication server.  But I'm not sure what happens in the background - even if it was downloading it on every restart - I don't see a problem as the content is small - and it would ensure the data is the most current.

I would assume that if I were to upgrade keycloak and there were changes in the results - I would simply need to restart geonetwork and it would pick up the new changes without having to manually apply the changes in the configuration.

Reply:
In either case I see that you added OPENIDCONNECT_SERVERMETADATA_CONFIG_URL.  Thank you - I will try to run my test again using that new option to see if I get any better luck.

Reply:
I tested it (with Azure) and it worked fine.  Should work fine with KeyCloak since its just downloading the JSON file.

KeyCloak and Geoserver allow users to either type in all the information about the OIDC server (into a web form) or you can download (and then edit by hand) the info from the metadata URL (it uses the JSON to fill in the web form).

The (perhaps hand edited results) are stored locally by Keycloak and GS.  They don't talk to the metadata server except in the UI.

Everyone on Azure can use the same JSON file - it just uses the client id to route. You can also use an azure-subscription-id metadata endpoint, but both work.

Let me know if it works for you.